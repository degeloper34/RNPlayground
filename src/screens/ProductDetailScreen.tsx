import {AntDesign, Ionicons, MaterialIcons} from "@expo/vector-icons";
import {useContext, useLayoutEffect, useState} from "react";
import {Image, Pressable, StyleSheet, View} from "react-native";
import {RootStackScreenProps} from "../../types";
import {CustomButton, CustomText, FlexView} from "../components/atoms";
import Colors from "../constants/Colors";
import {MainContext} from "../context/mainContext";

export default function ProductDetailScreen({
  navigation,
  route,
}: RootStackScreenProps<"ProductDetail">) {
  const product = route?.params?.product;
  const [quantity, setQuantity] = useState(1);
  const context = useContext(MainContext);

  useLayoutEffect(() => {
    navigation.setOptions({
      headerTitle: "",
      headerLeft: () => (
        <Pressable onPress={() => navigation.pop()}>
          <Ionicons name="arrow-back-outline" size={24} color={Colors.white} />
        </Pressable>
      ),
    });
  }, [navigation]);

  const onPressAddToCart = () => {
    context.updateCart(product, quantity, "add");
    navigation.pop();
  };

  const hideModal = () => {
    navigation.pop();
  };

  const onPressMinusQuantity = () => {
    if (quantity !== 1) {
      setQuantity(quantity - 1);
    }
  };

  const onPressPlusQuantity = () => {
    setQuantity(quantity + 1);
  };

  return (
    <View style={styles.container}>
      <Pressable
        onPress={hideModal}
        style={{flex: 1, backgroundColor: "rgba(0,0,0,0.5)"}}
      />
      <View
        style={{
          flex: 3,
          backgroundColor: Colors.mainNight,
          paddingHorizontal: 24,
          paddingTop: 24,
          paddingBottom: 36,
        }}
      >
        <View style={{flex: 1, backgroundColor: "white", borderRadius: 8}}>
          <Image
            source={{uri: product.image}}
            style={{flex: 1}}
            resizeMode={"contain"}
          />
        </View>
        <View style={{flex: 2, marginTop: 24}}>
          <CustomText
            text={product.title}
            type={"bold"}
            textColor={Colors.mainYellow}
            fontSize={14}
            numberOfLines={2}
          />
          <View style={{marginTop: 12}} />
          <CustomText
            text={product.description}
            type={"medium"}
            textColor={Colors.mainGrey}
            fontSize={12}
            numberOfLines={8}
          />

          <View style={{marginTop: 12}} />

          <CustomText
            text={"$" + product.price}
            type={"bold"}
            textColor={Colors.mainYellow}
            fontSize={14}
          />

          <CustomText
            text={"Please select quantity"}
            type={"bold"}
            textColor={Colors.white}
            fontSize={14}
            style={{marginTop: 24, marginBottom: 12}}
          />

          <View
            style={{
              flexDirection: "row",
              alignItems: "center",
            }}
          >
            <Pressable onPress={onPressMinusQuantity}>
              <AntDesign name="minuscircle" size={40} color={Colors.mainGrey} />
            </Pressable>

            <View
              style={{width: 40, marginHorizontal: 6, alignItems: "center"}}
            >
              <CustomText
                text={quantity}
                fontSize={18}
                type={"medium"}
                textColor={Colors.mainYellow}
              />
            </View>

            <Pressable onPress={onPressPlusQuantity}>
              <AntDesign name="pluscircle" size={40} color={Colors.mainGrey} />
            </Pressable>
          </View>
        </View>

        <View style={{flexDirection: "row", alignItems: "center"}}>
          <FlexView>
            <CustomButton text={"Add to Cart"} onPress={onPressAddToCart} />
          </FlexView>

          <Pressable onPress={hideModal} style={{marginLeft: 12}}>
            <MaterialIcons name="cancel" size={50} color={Colors.mainGrey} />
          </Pressable>
        </View>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
  },
});
