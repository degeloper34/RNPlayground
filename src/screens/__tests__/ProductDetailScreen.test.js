import renderer from "react-test-renderer";
import React from "react";
import ProductDetailScreen from "../ProductDetailScreen";

jest.mock("@expo/vector-icons");

describe("ProductDetailScreen", () => {
  it("should render without issues", () => {
    const setOptions = jest.fn();
    const navigate = jest.fn();
    const screen = renderer
      .create(
        <ProductDetailScreen
          navigation={{
            navigate,
            setOptions,
          }}
          route={{
            params: {
              product: {
                id: 1,
                title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
                category: "men's clothing",
                description:
                  "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
                image:
                  "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
                price: 109.95,
                rating: {
                  count: 120,
                  rate: 3.9,
                },
              },
            },
          }}
        />
      )
      .toJSON();

    expect(screen).toMatchSnapshot();
  });
});
