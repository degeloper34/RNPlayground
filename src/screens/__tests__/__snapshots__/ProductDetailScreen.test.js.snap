// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetailScreen should render without issues 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "rgba(0,0,0,0.5)",
        "flex": 1,
      }
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "#21325E",
        "flex": 3,
        "paddingBottom": 36,
        "paddingHorizontal": 24,
        "paddingTop": 24,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 8,
          "flex": 1,
        }
      }
    >
      <Image
        resizeMode="contain"
        source={
          Object {
            "uri": "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
          }
        }
        style={
          Object {
            "flex": 1,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "flex": 2,
          "marginTop": 24,
        }
      }
    >
      <Text
        allowFontScaling={false}
        numberOfLines={2}
        style={
          Array [
            undefined,
            Object {
              "color": "#F1D00A",
              "fontFamily": "sora-bold",
              "fontSize": 14,
              "textDecorationLine": "none",
            },
          ]
        }
      >
        Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
      </Text>
      <View
        style={
          Object {
            "marginTop": 12,
          }
        }
      />
      <Text
        allowFontScaling={false}
        numberOfLines={8}
        style={
          Array [
            undefined,
            Object {
              "color": "#F0F0F0",
              "fontFamily": "sora-medium",
              "fontSize": 12,
              "textDecorationLine": "none",
            },
          ]
        }
      >
        Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday
      </Text>
      <View
        style={
          Object {
            "marginTop": 12,
          }
        }
      />
      <Text
        allowFontScaling={false}
        style={
          Array [
            undefined,
            Object {
              "color": "#F1D00A",
              "fontFamily": "sora-bold",
              "fontSize": 14,
              "textDecorationLine": "none",
            },
          ]
        }
      >
        $109.95
      </Text>
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "marginBottom": 12,
              "marginTop": 24,
            },
            Object {
              "color": "#fff",
              "fontFamily": "sora-bold",
              "fontSize": 14,
              "textDecorationLine": "none",
            },
          ]
        }
      >
        Please select quantity
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "marginHorizontal": 6,
              "width": 40,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                undefined,
                Object {
                  "color": "#F1D00A",
                  "fontFamily": "sora-medium",
                  "fontSize": 18,
                  "textDecorationLine": "none",
                },
              ]
            }
          >
            1
          </Text>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        />
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#F1D00A",
              "borderRadius": 30,
              "justifyContent": "center",
              "padding": 12,
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                undefined,
                Object {
                  "color": "#21325E",
                  "fontFamily": "sora-bold",
                  "fontSize": 14,
                  "textDecorationLine": "none",
                },
              ]
            }
          >
            Add to Cart
          </Text>
        </View>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "marginLeft": 12,
          }
        }
      />
    </View>
  </View>
</View>
`;
